File: MechanicsCore/src/me/deecaad/core/file/Configuration.java
package me.deecaad.core.file;

import org.bukkit.configuration.ConfigurationSection;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.List;
import java.util.Set;
import java.util.function.BiConsumer;

public interface Configuration {

    /**
     * Adds every key and value from the given <code>ConfigurationSection</code>
     * to this configuration. If a key is added that already exists in
     * this <code>Configuration</code>, then a <code>DuplicateKeyException</code>
     * is thrown.
     *
     * @param config The configuration to pull from
     * @throws DuplicateKeyException If a duplicate key is found
     */
    void add(ConfigurationSection config) throws DuplicateKeyException;

    /**
     * Adds every key and value from the given <code>Configuration</code>
     * to this configuration. If a key is added that already exists in
     * this <code>Configuration</code>, then a <code>DuplicateKeyException</code>
     * is thrown.
     *
     * @param config The configuration to pull from
     * @throws DuplicateKeyException If a duplicate key is found
     */
    void add(Configuration config) throws DuplicateKeyException;

    /**
     * Sets the value at the given key
     *
     * @param key Location to set
     * @param value The value to set at the location
     * @return The value previously at that location, or null
     */
    @Nullable
    Object set(String key, Object value);

    /**
     * Gets every key present in this <code>Configuration</code>
     *
     * @return All keys
     */
    Set<String> getKeys();

    /**
     * Get the <code>int</code> value stored at
     * the given key
     *
     * @param key The location to pull the value from
     * @return The pulled value
     */
    int getInt(String key);

    /**
     * Get the <code>int</code> value stored at
     * the given key. If the key is not present,
     * the default value is returned.
     *
     * @param key The location to pull the value from
     * @param def The default value
     * @return The pulled value
     */
    int getInt(String key, int def);

    /**
     * Get the <code>double</code> value stored at
     * the given key
     *
     * @param key The location to pull the value from
     * @return The pulled value
     */
    double getDouble(String key);

    /**
     * Get the <code>Double</code> value stored at
     * the given key. If the key is not present,
     * the default value is returned.
     *
     * @param key The location to pull the value from
     * @param def The default value
     * @return The pulled value
     */
    double getDouble(String key, double def);

    /**
     * Get the <code>String</code> value stored at
     * the given key
     *
     * @param key The location to pull the value from
     * @return The pulled value
     */
    @Nonnull
    String getString(String key);

    /**
     * Get the <code>String</code> value stored at
     * the given key. If the key is not present,
     * the default value is returned.
     *
     * @param key The location to pull the value from
     * @param def The default value
     * @return The pulled value
     */
    String getString(String key, String def);

    /**
     * Get the <code>boolean</code> value stored at
     * the given key
     *
     * @param key The location to pull the value from
     * @return The pulled value
     */
    boolean getBool(String key);

    /**
     * Get the <code>boolean</code> value stored at
     * the given key. If the key is not present,
     * the default value is returned.
     *
     * @param key The location to pull the value from
     * @param def The default value
     * @return The pulled value
     */
    boolean getBool(String key, boolean def);

    /**
     * Get the <code>List</code> value stored at
     * the given key
     *
     * @param key The location to pull the value from
     * @return The pulled value
     */
    @Nonnull
    List<String> getList(String key);

    /**
     * Get the <code>List</code> value stored at
     * the given key. If the key is not present,
     * the default value is returned.
     *
     * @param key The location to pull the value from
     * @param def The default value
     * @return The pulled value
     */
    List<String> getList(String key, List<String> def);

    /**
     * Get the <code>Object</code> value stored at
     * the given key. If no object is found null is
     * returned
     *
     * @param key The location to pull the value from
     * @return The pulled value
     */
    @Nullable
    Object getObject(String key);

    /**
     * Get the <code>Object</code> value stored at
     * the given key. If the key is not present,
     * the default value is returned.
     *
     * @param key The location to pull the value from
     * @param def The default value
     * @return The pulled value
     */
    Object getObject(String key, Object def);

    /**
     * Get the <code>Object</code> value of type
     * <code>T</code> at the given key
     *
     * @param key The location to pull the value from
     * @param clazz The class used to cast to type T
     * @param <T> The data type of the value
     * @return The pulled and casted value
     */
    @Nullable
    <T> T getObject(String key, Class<T> clazz);

    /**
     * Get the <code>Object</code> value of type
     * <code>T</code> at the given key. If the key
     * is not present, the default value is returned.
     *
     * @param key The location to pull the value from
     * @param def The default value
     * @param clazz The class used to cast to type T
     * @param <T> The data type of the value
     * @return The pulled and casted value
     */
    <T> T getObject(String key, T def, Class<T> clazz);

    /**
     * Checks to see if the given key is contained within
     * this <code>Configuration</code>
     *
     * @param key The key to check for
     * @return true if the key is present
     */
    boolean containsKey(String key);

    /**
     * Checks to see if the given key is contained within
     * this <code>Configuration</code>. If the key is present,
     * then there is a second check to make sure it is of the
     * given data type
     *
     * @param key The key to check for
     * @param clazz The data type to check for
     * @return true if the key is present
     */
    boolean containsKey(String key, Class<?> clazz);

    /**
     * Clears this <code>Configuration</code> by removing
     * every key and value
     */
    void clear();

    /**
     * Loops through every key that starts with
     * <code>path</code> and checks to see if the
     * key is "deep" or not
     *
     * @param path The starting path
     * @param consumer What to do with every key
     * @param deep true if should go deep
     */
    void forEach(String path, BiConsumer<String, Object> consumer, boolean deep);
}

File: MechanicsCore/src/main/kotlin/me/deecaad/core/file/Configuration.kt
package me.deecaad.core.file

import kotlin.jvm.Throws

/**
 * Organizes all configuration files into some collection of key-value pairs,
 * where the key is always a string.
 */
abstract class Configuration : Iterable<Map.Entry<String, Any>> {
    // Mutators

    /**
     * Sets the value of the property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration, it is added.
     *
     * @param key the key of the property.
     * @param value the value of the property.
     */
    abstract fun set(
        key: String,
        value: Any?,
    )

    /**
     * Adds all properties from the specified [other] configuration to this configuration. If a key
     * from [other] is already present in this configuration, a [DuplicateKeyException] is thrown.
     *
     * @param other The other configuration to copy from.
     */
    @Throws(DuplicateKeyException::class)
    abstract fun copyFrom(other: Configuration)

    /**
     * Removes all key-value pairs from this configuration.
     */
    abstract fun clear()

    // Accessors

    /**
     * Returns an immutable set of all keys in this configuration.
     */
    abstract fun keys(): Set<String>

    /**
     * Returns an immutable set of all entries in this configuration.
     */
    abstract fun entries(): Set<Map.Entry<String, Any>>

    /**
     * Returns an immutable collection of all values in this configuration.
     */
    abstract fun values(): Collection<Any>

    /**
     * Returns true if this configuration contains the specified [key]. Otherwise, returns false.
     *
     * @param key the key to search for.
     * @return true if the key is present, false otherwise.
     */
    abstract operator fun contains(key: String): Boolean

    /**
     * Returns true if this configuration contains the specified [key], and that value is a boolean
     * value. Otherwise, returns false.
     */
    abstract fun hasBoolean(key: String): Boolean

    /**
     * Returns the value of the boolean property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasBoolean] returns false),
     * this method returns the specified default ([def]).
     *
     * @param key the key of the property.
     * @param def the default value, if the property is not present. Defaults to false.
     * @return the value of the property, or the default value
     */
    @JvmOverloads
    fun getBoolean(
        key: String,
        def: Boolean = false,
    ): Boolean {
        return getBoolean0(key, def)
    }

    // internal method to get the boolean value
    protected abstract fun getBoolean0(
        key: String,
        def: Boolean,
    ): Boolean

    /**
     * Returns true if this configuration contains the specified [key], and that value is an integer
     * value. Otherwise, returns false.
     */
    abstract fun hasInt(key: String): Boolean

    /**
     * Returns the value of the integer property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasInt] returns false),
     * this method returns the specified default ([def]).
     *
     * @param key the key of the property.
     * @param def the default value, if the property is not present. Defaults to 0.
     * @return the value of the property, or the default value
     */
    @JvmOverloads
    fun getInt(
        key: String,
        def: Int = 0,
    ): Int {
        return getInt0(key, def)
    }

    // internal method to get the integer value
    protected abstract fun getInt0(
        key: String,
        def: Int,
    ): Int

    /**
     * Returns true if this configuration contains the specified [key], and that value is a double
     * value. Otherwise, returns false.
     */
    abstract fun hasDouble(key: String): Boolean

    /**
     * Returns the value of the double property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasDouble] returns false),
     * this method returns the specified default ([def]).
     *
     * @param key the key of the property.
     * @param def the default value, if the property is not present. Defaults to 0.0.
     * @return the value of the property, or the default value
     */
    @JvmOverloads
    fun getDouble(
        key: String,
        def: Double = 0.0,
    ): Double {
        return getDouble0(key, def)
    }

    // internal method to get the double value
    protected abstract fun getDouble0(
        key: String,
        def: Double,
    ): Double

    /**
     * Returns true if this configuration contains the specified [key], and that value is an object
     * value. Otherwise, returns false.
     */
    abstract fun hasObject(key: String): Boolean

    /**
     * Returns the value of the object property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasObject] returns false),
     * this method returns null.
     *
     * @param key the key of the property.
     * @return the value of the property, or null
     */
    fun getObject(key: String): Any? {
        return getObject0(key, null, Any::class.java)
    }

    /**
     * Returns the value of the object property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasObject] returns false),
     * this method returns the specified default ([def]).
     *
     * @param key the key of the property.
     * @param def the default value, if the property is not present.
     * @return the value of the property, or the default value
     */
    fun getObject(
        key: String,
        def: Any,
    ): Any {
        return getObject0(key, def, Any::class.java)!!
    }

    /**
     * Returns the value of the object property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasObject] returns false),
     * this method returns null.
     *
     * @param key the key of the property.
     * @param clazz the class of the object to be returned.
     * @return the value of the property, or null
     */
    fun <T : Any> getObject(
        key: String,
        clazz: Class<T>,
    ): T? {
        return getObject0(key, null, clazz)
    }

    /**
     * Returns the value of the object property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasObject] returns false),
     * this method returns the specified default ([def]).
     *
     * @param key the key of the property.
     * @param def the default value, if the property is not present. Defaults to null.
     * @param clazz the class of the object to be returned.
     * @return the value of the property, or the default value
     */
    fun <T : Any> getObject(
        key: String,
        def: T,
        clazz: Class<T>,
    ): T {
        return getObject0(key, def, clazz)!!
    }

    /**
     * Returns the value of the object property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasObject] returns false),
     * this method returns null.
     *
     * @param key the key of the property.
     * @return the value of the property, or null
     */
    inline fun <reified T : Any> get(key: String): T? {
        return getObject(key, T::class.java)
    }

    /**
     * Returns the value of the object property with the specified [key] in this configuration.
     *
     * If the property is not present in this configuration (when [hasObject] returns false),
     * this method returns the specified default ([def]).
     *
     * @param key the key of the property.
     * @param def the default value, if the property is not present. Defaults to null.
     * @return the value of the property, or the default value
     */
    inline fun <reified T : Any> get(
        key: String,
        def: T,
    ): T {
        return getObject(key, def, T::class.java)
    }

    // internal method to get the object value
    protected abstract fun <T : Any> getObject0(
        key: String,
        def: T?,
        clazz: Class<T>,
    ): T?

    /**
     * Returns the value of the string property with the specified [key] in this configuration.
     * If the property is not present in this configuration, this method returns null.
     *
     * @param key the key of the property.
     * @return the value of the property, or null
     */
    fun getString(key: String): String? {
        return getObject(key, String::class.java)
    }

    /**
     * Returns the value of the string property with the specified [key] in this configuration.
     * If the property is not present in this configuration, this method returns the specified default ([def]).
     *
     * @param key the key of the property.
     * @param def the default value, if the property is not present.
     * @return the value of the property, or the default value
     */
    fun getString(
        key: String,
        def: String,
    ): String {
        return getObject(key, def, String::class.java)
    }
}
